akka-stream/src/main/scala/akka/stream/impl/SynchronousFileSubscriber.scala

```diff
@@ -0,0 +1,76 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl
+
+import java.io.{ File, RandomAccessFile }
+import java.nio.channels.FileChannel
+
+import akka.actor.{ ActorLogging, Props }
+import akka.stream.actor.{ ActorSubscriberMessage, OneByOneRequestStrategy }
+import akka.util.ByteString
+
+import scala.concurrent.Promise
+
+/** INTERNAL API */
+private[akka] object SynchronousFileSubscriber {
+  def props(f: File, completionPromise: Promise[Long], append: Boolean) =
+    Props(classOf[SynchronousFileSubscriber], f, completionPromise, append)
+
+}
+
+/** INTERNAL API */
+private[akka] class SynchronousFileSubscriber(f: File, completionPromise: Promise[Long], append: Boolean) extends akka.stream.actor.ActorSubscriber
+  with ActorLogging {
+
+  // TODO since ActorPublisher is typed, shouldn't ActorSubscriber be as well?
+
+  override protected val requestStrategy = OneByOneRequestStrategy // TODO better one here ?
+
+  private val fs = new RandomAccessFile(f, "rw") // best way to express this in JDK6, OpenOption are available since JDK7
+  private var chan: FileChannel = _
+
+  private var bytesWritten: Long = 0
+
+  override def preStart(): Unit = {
+    try {
+      chan = fs.getChannel
+
+      // manually supporting appending to files - in Java 7 we could use OpenModes: FileChannel.open(f, openOptions.asJava)
+      if (append) chan.position(chan.size())
+    } catch {
+      case ex: Exception ⇒
+        completionPromise.failure(ex)
```

Comment:
Too early, I think you can call `cancel()` here and when `OnSubscribe` is handled in aroundReceive it will be cancelled.
However, we must also stop the actor. Perhaps we should be more opinionated about the actor lifecycle of the ActorPublisher and ActorSubscriber, i.e. stop it when the stream is completed/cancelled/failed?